pymc_experimental/distributions/timeseries.py

```diff
@@ -0,0 +1,214 @@
+import warnings
+
+import numpy as np
+import pymc as pm
+import pytensor
+import pytensor.tensor as at
+from pymc.distributions.dist_math import check_parameters
+from pymc.distributions.distribution import (
+    Distribution,
+    SymbolicRandomVariable,
+    _moment,
+)
+from pymc.distributions.logprob import ignore_logprob, logp
+from pymc.distributions.shape_utils import (
+    _change_dist_size,
+    change_dist_size,
+    get_support_shape_1d,
+)
+from pymc.logprob.abstract import _logprob
+from pymc.util import check_dist_not_registered
+from pytensor.graph.basic import Node
+from pytensor.tensor import TensorVariable
+from pytensor.tensor.random.op import RandomVariable
+
+
+class DiscreteMarkovChainRV(SymbolicRandomVariable):
+    default_output = 1
+    _print_name = ("DiscreteMC", "\\operatorname{DiscreteMC}")
+
+    def update(self, node: Node):
+        return {node.inputs[-1]: node.outputs[0]}
+
+
+class DiscreteMarkovChain(Distribution):
+    r"""
+    A Discrete Markov Chain is a sequence of random variables
+    .. math::
+        \{x_t\}_{t=0}^T
+    Where transition probability P(x_t | x_{t-1}) depends only on the state of the system at x_{t-1}.
+
+    Parameters
+    ----------
+    P: tensor
+        Matrix of transition probabilities between states. Rows must sum to 1.
+        One of P or P_logits must be provided.
+    P_logit: tensor, optional
+        Matrix of tranisiton logits. Converted to probabilities via Softmax activation.
+        One of P or P_logits must be provided.
+    steps: tensor, optional
+        Length of the markov chain. Only needed if state is not provided.
+    init_dist : unnamed distribution, optional
+        Vector distribution for initial values. Unnamed refers to distributions
+        created with the ``.dist()`` API. Distribution should have shape n_states.
+        If not, it will be automatically resized. Defaults to pm.Categorical.dist(p=np.full(n_states, 1/n_states)).
+        .. warning:: init_dist will be cloned, rendering it independent of the one passed as input.
+    """
+
+    rv_type = DiscreteMarkovChainRV
+
+    def __new__(cls, *args, steps=None, **kwargs):
+        # Subtract 1 step to account for x0 given, better match user expectation of
+        # len(markov_chain) = steps
+        if steps is not None:
+            steps -= 1
```

Comment:
`support_shape_offset` will handle the cases where the time dimension is defined via size, but not if it is defined via steps. For example, `pm.AR.dist(rho=0.99, steps=10).eval().shape` gives `(11,)`, but  `pm.AR.dist(rho=0.99, shape=10).eval().shape` gives `(10,)`. This correction to steps unifies the output shape to `(10,)` in all cases.

Reply:
Why not make steps behave like other timeseries? A GRW with shape 10, has 9 steps. Most users will never see steps, since we advertise the shape/dims/observed API 

It is indeed 9 steps right?

Reply:
I have no idea what a "step" is I guess. It has 9 "state transitions", are those steps? I guess personally I interpreted "steps" to mean "time steps', which would be the length of the chain

Reply:
Anyway it's easy enough to remove this and just keep everything consistent.

Reply:
We need steps internally because we use size for RVs. Size does not contain info about support shape, which for univariate timeseries is the last one.

Reply:
Steps is the number of steps in the scan argument

Reply:
Of course. But then it gets x0 stuck onto it, which is the source of my confusion -- i wanted to adjust for that.

Reply:
The final distribution is now x0 plus the rest? I guess this made sense when x0 was not part of the distribution?

Reply:
Right, if x0 counts and I ask for 10 steps, i get back 11 -- 10 from the scan and 1 from x0. I thought if x0 is part of the distribution, it should scan for 1 less to account for the fact we already have the first "step".